### PR TITLE
Fix function rest type not being represented correctly

### LINKED ISSUE #1192: Assigning a vararg function to a function value whose type has a fixed number of args is failing
**Description:**
Both of the fallowing cases are failing
```ts
type F function(int, int) returns int;

public function main() {
    F f = foo;
    io:println(f);
}

function foo(int a, int b, int... c) returns int {
    return 0;
}
```
```ts
type F function(int, int, int) returns int;

public function main() {
    F f = foo;
    io:println(f);
}

function foo(int a, int b, int... c) returns int {
    return 0;
}
```

They should be valid because function types are contravariant in their parameter types. That is parameter type `[int, int, int...]` is a supertype of both `[int, int]` and `[int, int, int]`. Therefore function types `function(int ,int, int) returns int` and `function(int, int) returns int` must be super types of `function(int, int, ...int) returns int`

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Fix function rest type not being represented correctly
Resolves #1192 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/front/expr.bal [hunk 1]
```diff
@@ -1835,8 +1835,20 @@ function functionRefTy(t:Context tc, bir:FunctionRef value) returns t:SemType {
     t:Env env = tc.env;
```
Comment (jclark) [inline compiler/modules/front/expr.bal]: functionRefSemType

Comment (heshanpadmasiri) [reply compiler/modules/front/expr.bal]: fixed

**Review by jclark — DISMISSED** [review]

**Comment by jclark** [comment]
Fix up subset15.md in a separate pull request

File: compiler/modules/types/sexprParse.bal [hunk 2]
```diff
@@ -204,15 +204,24 @@ function fromAtomSexpr(SexprAtomParseContext pc, string name, ts:Atom atomSexpr)
...
-            FunctionDefinition d = new;
-            pc.started[name] = d;
-            ListDefinition listDef = new;
-            SemType[] initial = from var member in <ts:Type[]>args select fromSexprInternal(pc, member);
-            SemType argsTuple = defineListTypeWrapped(listDef, pc.env, initial, initial.length(), mut = CELL_MUT_NONE);
-            return d.define(pc.env, argsTuple, fromSexprInternal(pc, <ts:Type>ret));
+            return fromMappingSexpr(pc, name, <ts:Field[]>fieldsSexpr, <ts:Type>rest);
+        }
+        ["function", var ret, var params] => {
+            readonly & SemType[] paramTypes = from var param in <ts:Type[]>params select fromSexprInternal(pc, param);
```
Comment (jclark) [inline compiler/modules/types/sexprParse.bal]: Factor this out into a function.

Comment (heshanpadmasiri) [reply compiler/modules/types/sexprParse.bal]: Fixed

**Review by jclark — APPROVED** [review]

**Pull request merged**